enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,948 @@
+`NetworkSelectionElements` must feature the owner information required by
+the OVN-Kubernetes CNI to use in the owner references via CNI args.
+
+Check the following network selection elements annotation from a KubeVirt
+launcher pod to see an example:
+
+```bash
+kubectl get pods <launcher pod name> -ojsonpath="{ @.metadata.annotations.k8s\.v1\.cni\.cncf\.io\/networks }" | jq
+[
+  {
+    "name": "tenantred",
+    "namespace": "default",
+    "interface": "pod16367aacb67",
+    "cni-args": {
+      "persistentip.cni.cncf.io/ownerID": "a0790345-4e84-4257-837a-e3d762d191ab",
+      "persistentip.cni.cncf.io/ownerName": "vm-a",
+      "persistentip.cni.cncf.io/ownerType": "VirtualMachine",
+      "persistentip.cni.cncf.io/ownerVersion": "kubevirt.io/v1"
+    }
+  }
+]
+```
+
+#### KubeVirt feature gate
+We recommend to protect this feature behind a feature gate, at least in the
+beginning, since the current proposal will not work safely for VM foreground
+deletes - i.e. the feature comes with a cost: only background delete
+propagation is accepted.
+
+More details about this in
+[Kubernetes GC extended info](#kubernetes-garbage-collection-extended-info).
+
+#### KubeVirt API changes
+
+The persistent IP allocations represented by the `IPAMLease` objects should be
+reported in the VM status, allowing the user to easily identify which IPs are
+pinned for the VM.
+
+Thus, we propose the adding the following struct to the
+[VirtualMachineStatus](https://github.com/kubevirt/kubevirt/blob/960484df024ae8cce0f189d5f492667c4d57b250/staging/src/kubevirt.io/api/core/v1/types.go#L1486)
+structure:
+
+```go
+// VirtualMachineStatus represents the status returned by the
+// controller to describe how the VirtualMachine is doing
+type VirtualMachineStatus struct {
+	...
+	NetworkInterfacesStatus []NetworkInterfaceStatus `json:"networkInterfacesStatus,omitempty"`
+}
+
+type NetworkInterfaceStatus {
+	InterfaceName string   `json:"interfaceName,omitempty"`
+	IPs           []string `json:"ips,omitempty"`
+}
+```
+
+**NOTE:** this change is not required per-se. It does provide helpful
+information to the VM user.
+
+### OVN-Kubernetes related changes
+When the OVN-Kubernetes `ovnkube-control-plane` pod realizes the network
+selection element features those CNI arguments, it will generate an IP address
+allocation for it, using the existing allocator, plus create an `IPAMLease` CR
+to represent it in the Kubernetes data-store. OVN-Kubernetes will use the
+CNI-args information to fill out the `ownerReferences` metadata on the
+`IPAMLease` CR, which would look like:
+
+```yaml
+apiVersion: "persistentip.cni.cncf.io/v1alpha1"
+kind: IPAMLease
+metadata:
+  name: vm-a.tenantred.pod16367aacb67
+  namespace: ns1
+  ownerReferences:
+  - apiVersion: kubevirt.io/v1
+    kind: VirtualMachine
+    name: vm-a
+    uid: a0790345-4e84-4257-837a-e3d762d191ab
+spec:
+  network: tenantred
+  interface: pod16367aacb67
+  ips:
+  - 10.128.20.8/24
+  - fd10:128:20::8/64
```

Comment:
Status

Reply:
discussed previouly; IMHO we should understand who creates the IPAM claim, since that does impact the need for a status subresource.